File: day9/puzzle1.dart
import 'dart:convert';
import 'dart:io';
import 'dart:math';

/// Extrapolates the last value of the specified [List].
int extrapolate(List<int> values) {
  var sequences = <List<int>>[];
  var sequence = values;
  while (true) {
    sequences.add(sequence);
    if ((sequence.reduce(max) == 0) && (sequence.reduce(min) == 0)) {
      break;
    }
    var differences = <int>[];
    for (var i = 0; i < sequence.length - 1; i++) {
      differences.add(sequence[i + 1] - sequence[i]);
    }
    sequence = differences;
  }
  int last = 0;
  for (var i = sequences.length - 1; i >= 0; i--) {
    sequence = sequences[i];
    if (i == sequences.length - 1) {
      sequence.add(last);
    } else {
      last = sequence.last + last;
      sequence.add(last);
    }
  }
  return last;
}

/// Application entry-point.
void main() async {
  final lines = utf8.decoder
      .bind(File("puzzle_input.txt").openRead())
      .transform(const LineSplitter());
  int sum = 0;
  await for (final line in lines) {
    sum += extrapolate([for (final value in line.split(' ')) int.parse(value)]);
  }
  print('sum = $sum');
}

File: day9/puzzle1.py
def extrapolate(values: list[int]) -> int:
    """Extrapolates the next value.

    Parameters
    ----------
    values : list[int]
        The values.

    Returns
    -------
    int
        The extrapolated value.
    """
    sequences: list[list[int]] = []
    sequence = values
    while True:
        sequences.append(sequence)
        if min(sequence) == 0 and max(sequence) == 0:
            break
        differences = []
        for i in range(len(sequence) - 1):
            difference = sequence[i + 1] - sequence[i]
            differences.append(difference)
        sequence = differences
    last = 0
    for i in range(len(sequences) - 1, -1, -1):
        sequence = sequences[i]
        if i == len(sequences) - 1:
            sequence.append(last)
        else:
            last = sequence[-1] + last
            sequence.append(last)
    return last


def main():
    """Application entry-point."""
    sum = 0
    with open("puzzle_input.txt", "rt") as f:
        for line in f:
            values = [int(value) for value in line.split(" ")]
            sum += extrapolate(values)
    print(f"sum = {sum}")


if __name__ == "__main__":
    main()
